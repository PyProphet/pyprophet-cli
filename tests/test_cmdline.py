# encoding: utf-8
from __future__ import print_function
import pdb
import subprocess
import pytest
import os
import shutil
import collections
import tempfile

import pandas

pandas.options.display.precision=6
pandas.options.display.width=200


@pytest.fixture(scope="session")
def test_data_folder():
    here = os.path.dirname(os.path.abspath(__file__))
    return os.path.join(here, "data")


@pytest.fixture(scope="session", params=[1, 3])
def setup(test_data_folder, request):
    num_files = request.param
    tmpdir = tempfile.mkdtemp()
    data_folder = os.path.join(tmpdir, "data_folder")
    os.makedirs(data_folder)
    working_folder = os.path.join(tmpdir, "working_folder")
    for i in range(num_files):
        shutil.copy(os.path.join(test_data_folder, "test_data.txt"), os.path.join(data_folder, "data_%d.txt" % i))

    Setup = collections.namedtuple("Setup", "data_folder working_folder number_input_files")
    return Setup(data_folder, working_folder, num_files)


def test_check(setup):
    cmd = ("pyprophet-brutus check --data-folder %s" % setup.data_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0


def test_subsample(setup):
    cmd = ("pyprophet-brutus subsample --job-number 1 --job-count 1 "
           "--sample-factor 0.1 "
           "--random-seed 43 "
           "--ignore-invalid-scores "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --working-folder %s") % (setup.data_folder, setup.working_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
    files = os.listdir(setup.working_folder)
    assert len(files) == setup.number_input_files + 1

    files.sort()
    subsamples = pandas.read_csv(os.path.join(setup.working_folder, files[-1]), sep="\t")
    assert subsamples.shape == (934, 21)

    cmd = ("pyprophet-brutus subsample --job-number 1 --job-count 2 "
           "--sample-factor 0.1 "
           "--random-seed 43 "
           "--ignore-invalid-scores "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --working-folder %s "
           "--local-folder %s") % (setup.data_folder, setup.working_folder, tempfile.mkdtemp())
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
    files = os.listdir(setup.working_folder)
    assert len(files) == 2 if setup.number_input_files == 1 else 3

    files.sort()
    subsamples = pandas.read_csv(os.path.join(setup.working_folder, files[-1]), sep="\t")
    assert subsamples.shape == (934, 21)

    cmd = ("pyprophet-brutus subsample --job-number 1 --job-count 2 "
           "--sample-factor 0.1 "
           "--random-seed 43 "
           "--ignore-invalid-scores "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --working-folder %s") % (setup.data_folder, setup.working_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
    files = os.listdir(setup.working_folder)
    assert len(files) == 2 if setup.number_input_files == 1 else 3

    files.sort()

    subsamples = pandas.read_csv(os.path.join(setup.working_folder, files[-1]), sep="\t")
    assert subsamples.shape == (934, 21)


def test_learn(setup, regtest):
    cmd = ("pyprophet-brutus learn --random-seed 43 --working-folder %s" % setup.working_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0

    files = sorted(os.listdir(setup.working_folder))

    for i, file_ in enumerate(files):
        path = os.path.join(setup.working_folder, file_)
        print(i, "%7d" % os.stat(path).st_size, file_, file=regtest)

    for name in ("weights.txt", "sum_stat_subsampled.txt"):
        df = pandas.read_csv(os.path.join(setup.working_folder, name), header=None, sep="\t")
        print(df, file=regtest)
        # print(open(os.path.join(setup.working_folder, name), "r").read(), file=regtest)


def test_score(setup, regtest):
    cmd = ("pyprophet-brutus apply_weights --job-number 1 --job-count 1 "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --working-folder %s") % (setup.data_folder, setup.working_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
