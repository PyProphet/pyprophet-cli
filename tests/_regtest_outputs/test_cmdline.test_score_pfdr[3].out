 0 2826133 result_folder_3/data_0_scored.txt
 1 2826133 result_folder_3/data_1_scored.txt
 2 2826133 result_folder_3/data_2_scored.txt
 3   29726 result_folder_3/report.pdf
 4   29726 result_folder_3/report_grouped_by_transition_group_id.pdf
 5    1165 result_folder_3/summary_stats.txt
 6    1165 result_folder_3/summary_stats_grouped_by_transition_group_id.txt
 0   89146 work_folder/data_0_score_data.npz
 1 2162728 work_folder/data_0_top_score_data.hdf
 2   89146 work_folder/data_1_score_data.npz
 3 2162728 work_folder/data_1_top_score_data.hdf
 4   89146 work_folder/data_2_score_data.npz
 5 2162728 work_folder/data_2_top_score_data.hdf
 6      20 work_folder/extra_group_columns.txt
 7    6531 work_folder/full_stat_subsampled.txt
 8      17 work_folder/invalid_columns.txt
 9     392 work_folder/score_columns.txt
10 1049943 work_folder/subsampled_data_0.txt
11 1049943 work_folder/subsampled_data_1.txt
12 1049943 work_folder/subsampled_data_2.txt
13     708 work_folder/sum_stat_subsampled.txt
14     433 work_folder/weights.txt

num_null   : 25.00
num_total  : 387.00
stats shape: (387, 10)
mean top target scores: 4.367879
sdev top target scores: 1.518136
mean top decoy  scores: -0.000000
sdev top decoy  scores: 1.000000

   qvalue    svalue    pvalue   TP  FP            TN            FN       FDR      sens    cutoff
0    0.00  0.872915  0.000186  316   0  2.500000e+01  4.600000e+01  0.000212  0.872928  3.559014
1    0.01  0.998093  0.147618  361   4  2.100000e+01  1.000000e+00  0.010111  0.997238  1.046706
2    0.02  1.000000  0.282889  361   7  1.800000e+01  1.000000e+00  0.019218  0.997238  0.574279
3    0.05  1.000000  0.795019  363  20  5.000000e+00 -1.000000e+00  0.051894  1.000000 -0.823961
4    0.10  1.000000  0.982481  362  25 -3.552714e-15  3.552714e-15  0.063468  1.000000 -2.107914
5    0.20       NaN       NaN  NaN NaN           NaN           NaN       NaN       NaN       NaN
6    0.30       NaN       NaN  NaN NaN           NaN           NaN       NaN       NaN       NaN
7    0.40       NaN       NaN  NaN NaN           NaN           NaN       NaN       NaN       NaN
8    0.50       NaN       NaN  NaN NaN           NaN           NaN       NaN       NaN       NaN

num_null   : 25.00
num_total  : 387.00
stats shape: (387, 10)
mean top target scores: 4.367879
sdev top target scores: 1.518136
mean top decoy  scores: -0.000000
sdev top decoy  scores: 1.000000

   qvalue    svalue    pvalue   TP  FP            TN            FN       FDR      sens    cutoff
0    0.00  0.872915  0.000186  316   0  2.500000e+01  4.600000e+01  0.000212  0.872928  3.559014
1    0.01  0.998093  0.147618  361   4  2.100000e+01  1.000000e+00  0.010111  0.997238  1.046706
2    0.02  1.000000  0.282889  361   7  1.800000e+01  1.000000e+00  0.019218  0.997238  0.574279
3    0.05  1.000000  0.795019  363  20  5.000000e+00 -1.000000e+00  0.051894  1.000000 -0.823961
4    0.10  1.000000  0.982481  362  25 -3.552714e-15  3.552714e-15  0.063468  1.000000 -2.107914
5    0.20       NaN       NaN  NaN NaN           NaN           NaN       NaN       NaN       NaN
6    0.30       NaN       NaN  NaN NaN           NaN           NaN       NaN       NaN       NaN
7    0.40       NaN       NaN  NaN NaN           NaN           NaN       NaN       NaN       NaN
8    0.50       NaN       NaN  NaN NaN           NaN           NaN       NaN       NaN       NaN

transition_group_id	run_id	decoy	main_var_xx_swath_prelim_score	var_bseries_score	var_elution_model_fit_score	var_intensity_score	var_isotope_correlation_score	var_isotope_overlap_score	var_library_corr	var_library_rmsd	var_log_sn_score	var_massdev_score	var_massdev_score_weighted	var_norm_rt_score	var_xcorr_coelution	var_xcorr_coelution_weighted	var_xcorr_shape	var_xcorr_shape_weighted	var_yseries_score	var_faulty_score	peak_group_rank	d_score	m_score	p_value	transition_group_id_m_score	transition_group_id_p_value

459_run0	0	0	5.27893972397	5.0	0.993098139763	0.930480718613	0.998724222183	0.0	0.912406027317	0.0398625209928	5.01656723022	8.83918857574	3.88453722	0.00229200161994	0.0	0.0	0.960350096226	0.970979332924	6.0		1	5.00816202164	0.000211860635318	2.74761021046e-07	0.000211860635318	2.74761021046e-07


transition_group_id	run_id	decoy	main_var_xx_swath_prelim_score	var_bseries_score	var_elution_model_fit_score	var_intensity_score	var_isotope_correlation_score	var_isotope_overlap_score	var_library_corr	var_library_rmsd	var_log_sn_score	var_massdev_score	var_massdev_score_weighted	var_norm_rt_score	var_xcorr_coelution	var_xcorr_coelution_weighted	var_xcorr_shape	var_xcorr_shape_weighted	var_yseries_score	var_faulty_score	peak_group_rank	d_score	m_score	p_value	transition_group_id_m_score	transition_group_id_p_value

459_run0	0	0	5.27893972397	5.0	0.993098139763	0.930480718613	0.998724222183	0.0	0.912406027317	0.0398625209928	5.01656723022	8.83918857574	3.88453722	0.00229200161994	0.0	0.0	0.960350096226	0.970979332924	6.0		1	5.00816202164	0.000211860635318	2.74761021046e-07	0.000211860635318	2.74761021046e-07


transition_group_id	run_id	decoy	main_var_xx_swath_prelim_score	var_bseries_score	var_elution_model_fit_score	var_intensity_score	var_isotope_correlation_score	var_isotope_overlap_score	var_library_corr	var_library_rmsd	var_log_sn_score	var_massdev_score	var_massdev_score_weighted	var_norm_rt_score	var_xcorr_coelution	var_xcorr_coelution_weighted	var_xcorr_shape	var_xcorr_shape_weighted	var_yseries_score	var_faulty_score	peak_group_rank	d_score	m_score	p_value	transition_group_id_m_score	transition_group_id_p_value

459_run0	0	0	5.27893972397	5.0	0.993098139763	0.930480718613	0.998724222183	0.0	0.912406027317	0.0398625209928	5.01656723022	8.83918857574	3.88453722	0.00229200161994	0.0	0.0	0.960350096226	0.970979332924	6.0		1	5.00816202164	0.000211860635318	2.74761021046e-07	0.000211860635318	2.74761021046e-07

